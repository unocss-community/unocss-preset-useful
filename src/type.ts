export interface UsefulOptions {
  /**
   * Extract rgba color in css variable
   *
   * @default false
   */
  unColor?: boolean | string

  /**
   * Expand theme animation name usage
   *
   * [ name, duration, timing-function, iteration-count ]
   *
   * @example
   * ```ts
    themeAnimate: ['spin 1s linear infinite'],
   * ```
   */
  themeAnimate?: string[]
}

export type ResolvedOptions = Required<UsefulOptions>

export type DeepPartial<T> = { [P in keyof T]?: DeepPartial<T[P]> }
