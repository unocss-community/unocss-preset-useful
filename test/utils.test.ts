import type { CSSObject } from 'unocss'
import { describe, expect, test } from 'vitest'
import { cssObj2StrSync, nomarlizeAnimate, stringifyObj } from 'unocss-preset-useful'

describe('utils', () => {
  const animate = [
    'spin 1s linear infinite',
    'bounce 2s ease-in-out 3',
    'fade 1s ease-in-out 3',
    'foo 1s * 3',
    'bar 1s +',
  ]

  test('nomarlizeAnimate', async () => {
    const result = nomarlizeAnimate({ animate })

    expect(result).toMatchSnapshot()
  })

  test('cssObj2StrSync', async () => {
    const style: Record<string, CSSObject> = {
      '0%, 20%, 40%, 50%': { opacity: 1, transform: 'scale(1, 1) translateY(0)' },
      '10%, 30%': { opacity: 1, transform: 'scale(1.1, 1.1) translateY(0)' },
      '100%': { opacity: 0, transform: 'scale(1, 1) translateY(-900%)' },
    }

    expect(await cssObj2StrSync(style)).toMatchInlineSnapshot('"0%, 20%, 40%, 50%{opacity:1;transform:scale(1, 1) translateY(0);}10%, 30%{opacity:1;transform:scale(1.1, 1.1) translateY(0);}100%{opacity:0;transform:scale(1, 1) translateY(-900%);}"')
  })

  test('stringify CSSObject', () => {
    const obj = {
      foo: 'bb',
      bar: 'dd',
    }

    expect(stringifyObj(obj)).toMatchInlineSnapshot('"{foo:bb;bar:dd;}"')
  })
})
