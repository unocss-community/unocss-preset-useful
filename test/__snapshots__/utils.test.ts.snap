// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`utils > themeAnimate configuration 1`] = `
{
  "counts": {
    "bounce": "3",
    "fade": "3",
    "spin": "infinite",
    "test": "3",
  },
  "durations": {
    "aaa": "1s",
    "bounce": "2s",
    "fade": "1s",
    "spin": "1s",
    "test": "1s",
  },
  "timingFns": {
    "aaa": "",
    "bounce": "ease-in-out",
    "fade": "ease-in-out",
    "spin": "linear",
  },
}
`;
